File: Dynamic Programming/509. Fibonacci Number/509. Fibonacci Number.js
/**
 * @param {number} n
 * @return {number}
 */
var fib = function (n) {
  if (n < 2) return n;

  let prev = 0;
  let curr = 1;

  for (let i = 2; i <= n; i++) {
    [prev, curr] = [curr, prev + curr];
  }

  return curr;
};

// Time: O(n)
// Space: O(1)

File: Dynamic Programming/509. Fibonacci Number/509. Fibonacci Number.py
class Solution:
    def fib(self, n: int) -> int:
        if n < 2:
            return n

        prev, curr = 0, 1
        for _ in range(2, n + 1):
            prev, curr = curr, curr + prev
        
        return curr

# Time: O(n)
# Space: O(1)